### PR TITLE
Configurable support for PROXY UNKNOWN

### PR DESCRIPTION
**Pull request by jehiah**
This adds a configurable way to enable support for `PROXY UNKNOWN` by setting `&Listener{..., UnknownOK: true}`

This support is useful for handling requests over the PROXY protocol when the request doesn't have a source IP address to set. An example of when that happens is GCP TCP Load Balancing Health checks.

https://cloud.google.com/load-balancing/docs/tcp/setting-up-tcp#proxy-protocol

### CODE REVIEW

**Comment by jehiah** [comment]
Hi @armon Is there anything i can do here to help get this merged?

**Comment by armon** [comment]
@jehiah Unfortunately I don't actively code in my role any more. This library is used by HashiCorp products, but we don't have a team who owns it at the moment, so it is in limbo. I'd recommend using your patch from a fork, and we are working on spinning up a team to take ownership of our shared libraries. Hopefully that will help unblock the issues / PRs on our libraries that are not core HashiCorp products.

File: protocol.go
Comment (jefferai) [inline protocol.go]: I don't believe this needs to be exported

Comment (jehiah) [reply protocol.go]: good catch. Un-exported.

**Comment by jehiah** [comment]
Thanks @armon and @jefferai I appreciate the follow up. I certainly didn't wait for this to land before starting to use this, but I'd love to see it land upstream if possible.

**Comment by jehiah** [comment]
@armon and @jefferai I'd love to follow up on this and see if we can get this merged. Is there anything else i can do to help?

In case it helps, I have been using this reliably in a production environment for a while now.

**Comment by jefferai** [comment]
I think it looks good. One request: can you slightly change the error message to disambiguate it? E.g. "Invalid UNKNOWN header field" or something.

**Comment by jehiah** [comment]
@jefferai good suggestion; I've updated that error message, and cleaned up the commits on this branch so it should be ready for merge.

**Review by jefferai — APPROVED** [review]

**Comment by jefferai** [comment]
Looks good. Thanks!

**Pull request merged**
